src/coreclr/jit/lowerriscv64.cpp

```diff
@@ -0,0 +1,880 @@
+XX                                                                           XX
+XX             Lowering for RISCV64 common code                              XX
+XX                                                                           XX
+XX  This encapsulates common logic for lowering trees for the RISCV64        XX
+XX  architectures.  For a more detailed view of what is lowering, please     XX
+XX  take a look at Lower.cpp                                                 XX
+XX                                                                           XX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+*/
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#ifdef TARGET_RISCV64 // This file is ONLY used for RISCV64 architectures
+
+#include "jit.h"
+#include "sideeffects.h"
+#include "lower.h"
+#include "lsra.h"
+
+#ifdef FEATURE_HW_INTRINSICS
+#include "hwintrinsic.h"
+#endif
+
+//------------------------------------------------------------------------
+// IsCallTargetInRange: Can a call target address be encoded in-place?
+//
+// Return Value:
+//    True if the addr fits into the range.
+//
+bool Lowering::IsCallTargetInRange(void* addr)
+{
+    // TODO-RISCV64: using B/BL for optimization.
+    return false;
+}
+
+//------------------------------------------------------------------------
+// IsContainableImmed: Is an immediate encodable in-place?
+//
+// Return Value:
+//    True if the immediate can be folded into an instruction,
+//    for example small enough and non-relocatable.
+//
+bool Lowering::IsContainableImmed(GenTree* parentNode, GenTree* childNode) const
+{
+    if (!varTypeIsFloating(parentNode->TypeGet()))
+    {
+        // Make sure we have an actual immediate
+        if (!childNode->IsCnsIntOrI())
+            return false;
+        if (childNode->AsIntCon()->ImmedValNeedsReloc(comp))
+            return false;
+
+        // TODO-CrossBitness: we wouldn't need the cast below if GenTreeIntCon::gtIconVal had target_ssize_t type.
+        target_ssize_t immVal = (target_ssize_t)childNode->AsIntCon()->gtIconVal;
+
+        switch (parentNode->OperGet())
+        {
+            case GT_CMPXCHG:
+            case GT_LOCKADD:
+            case GT_XADD:
+                NYI_RISCV64("GT_CMPXCHG,GT_LOCKADD,GT_XADD");
+                break;
+
+            case GT_ADD:
+            case GT_EQ:
+            case GT_NE:
+            case GT_LT:
+            case GT_LE:
+            case GT_GE:
+            case GT_GT:
+                return emitter::isValidSimm12(immVal);
+            case GT_AND:
+            case GT_OR:
+            case GT_XOR:
+                return emitter::isValidUimm11(immVal);
+            case GT_JCMP:
+                assert(((parentNode->gtFlags & GTF_JCMP_TST) == 0) ? (immVal == 0) : isPow2(immVal));
```

Comment:
Does RISC-V have the equivalent of ARM64's `tbnz`/`tbz` instructions?

Reply:
Nope. In general, replace `TNBZ Rt, #imm, label` with  `SLLI tmp, Rt, 63-imm; BLTZ tmp, label` and TBZ with the same but BGEZ. 2 byte instruction plus 4 byte branch instruction.

For imm in 0..10 you could use ANDI with 1<<imm, then BEQZ or BNEZ, which is a 4 byte instruction then a 2 byte branch...  But it's the same size and performance anyway

Reply:
@BruceForstall Thank you so much.

Reply:
@brucehoult , that is :)

@brucehoult -- looks like you have lots of RISC-V experience. Maybe we'll see you contribute to the .NET implementation?

Reply:
> @brucehoult -- looks like you have lots of RISC-V experience.

Since you asked :-) I worked on the CoreCLR to Arm port at Samsung R&D Moscow in 2016, bought one of the first HIFive1 RISC-V boards in Jan 2017, then in early 2018 left Samsung to join RISC-V pioneers SiFive in San Mateo where I worked on runtime, compiler, emulator stuff and also contributed to designing the B and V extensions. (and less so others). You'll find my name in the credits for those extensions, plus the unprivileged ISA manual. Also co-moderator of /r/riscv on Reddit since December 2019, in which time we've grown it from 2500 to 14000 members.

Today was my last day at a New Zealand web dev company I've been working at during COVID. Looking at moving back into the RISC-V world now it's starting to explode. Have VisionFive 2 :-)  Currently have ssh access to a 64 core 2.0 GHz OoO (similar to Cortex A72) SG2042 and doing a little vector 0.7.1 asm work on that for someone ... mmmm, tasty.

Reply:
Thanks @brucehoult! Good luck in the RISC-V world :)

@clamp03 If there is no direct equivalent of `tbnz`/`tbz` then RISC-V should not be setting this `GTF_JCMP_TST` flag. Indeed, I can see that it is never set by `LowerJTrue`, so this part of the asset here is just dead code.

Reply:
I removed the assert code in the last patch `https://github.com/dotnet/runtime/pull/82379/commits/607dd64ec3dbefa2115317fccc95a6a3df483a14`